src/order/closure.lean

```diff
@@ -112,12 +127,45 @@ set.ext $ λ x, ⟨λ h, ⟨x, h⟩, by { rintro ⟨y, rfl⟩, apply c.idempoten
 /-- Send an `x` to an element of the set of closed elements (by taking the closure). -/
 def to_closed (x : α) : c.closed := ⟨c x, c.closure_is_closed x⟩
 
+lemma mk₃.to_closed {f : α → α} {p : α → Prop} {hf : ∀ x, x ≤ f x} {hfp : ∀ x, p (f x)}
+  {hmin : ∀ ⦃x y⦄, x ≤ y → p y → f x ≤ y} {x : α} {hx : p x} :
+  (mk₃ f p hf hfp hmin).closed x :=
+le_antisymm (hmin (le_refl _) hx) (hf _)
+
 lemma top_mem_closed {α : Type u} [order_top α] (c : closure_operator α) : ⊤ ∈ c.closed :=
 c.closure_top
 
 lemma closure_le_closed_iff_le {x y : α} (hy : c.closed y) : x ≤ y ↔ c x ≤ y :=
 by rw [← c.closure_eq_self_of_mem_closed hy, le_closure_iff]
 
+lemma closure_closure_sup_left {α : Type u} [semilattice_sup α] (c : closure_operator α)
+  (x y : α) :
+  c (c x ⊔ y) = c (x ⊔ y) :=
+le_antisymm ((closure_le_closed_iff_le c (closure_is_closed _ _)).1 (sup_le (c.monotone le_sup_left)
+  (le_trans le_sup_right (le_closure _ _)))) (c.monotone (sup_le_sup_right (le_closure _ _) _))
+
+lemma closure_closure_sup_right {α : Type u} [semilattice_sup α] (c : closure_operator α)
+  (x y : α) :
+  c (x ⊔ c y) = c (x ⊔ y) :=
+by { rw [sup_comm, closure_closure_sup_left, sup_comm] }
+
+lemma closure_closure_sup {α : Type u} [semilattice_sup α] (c : closure_operator α)
+  (x y : α) :
+  c (c x ⊔ c y) = c (x ⊔ y) :=
+by { rw [closure_closure_sup_left, closure_closure_sup_right] }
+
+lemma closure_closure_supr {α : Type u} {ι : Sort v} [complete_lattice α] (c : closure_operator α)
+  (x : ι → α) :
+  c (⨆ i, c (x i)) = c (⨆ i, x i) :=
+le_antisymm ((closure_le_closed_iff_le c (closure_is_closed _ _)).1 (supr_le (λ i, c.monotone
+  (le_supr _ _)))) (c.monotone (supr_le_supr (λ i, c.le_closure _)))
+
+lemma closure_closure_bsupr {α : Type u} [complete_lattice α] (c : closure_operator α)
+  (p : α → Prop) :
+  c (⨆ x (h : p x), c x) = c (⨆ x (h : p x), x) :=
+le_antisymm ((closure_le_closed_iff_le c (closure_is_closed _ _)).1 (bsupr_le (λ x hx, c.monotone
+  (le_bsupr_of_le x hx (le_refl x))))) (c.monotone (bsupr_le_bsupr (λ x hx, le_closure _ _)))
```

Comment:
I think the following renames make the meaning clearer:

* `closure_closure_sup_left` → `closure_sup_closure_left`
* `closure_closure_sup_right` → `closure_sup_closure_right`
* `closure_closure_sup` → `closure_sup_closure`
* `closure_closure_supr` → `closure_supr_closure`
* `closure_closure_bsupr` → `closure_bsupr_closure`

Since then the order of the word matches the nesting of the operations